src/hardware/sblaster.cpp

```diff
@@ -903,10 +943,54 @@ static uint32_t read_dma_16bit(const uint32_t bytes_to_read, const uint32_t i =
 	return check_cast<uint32_t>(bytes_read);
 }
 
-static void enqueue_frames(std::unique_ptr<AudioVector> frames)
+static void enqueue_frames(std::vector<AudioFrame>& frames)
 {
-	frames_added_this_tick += frames->num_frames;
-	soundblaster_mixer_queue.NonblockingEnqueue(std::move(frames));
+	soundblaster_mixer_queue.NonblockingBulkEnqueue(frames);
+	frames_added_this_tick += check_cast<int>(frames.size());
+}
+
+// Stretch the given number of mono DAC samples to the requested number of frames
+static std::vector<AudioFrame>& stretch_dac_to_frames(const int16_t* dac_samples,
+                                                      const int num_dac_samples,
+                                                      const int frames_requested,
+                                                      const ResampleMethod resample_method)
+{
+	assert(dac_samples && num_dac_samples > 0 && frames_requested > 0);
+
+	const auto dac_end_index    = num_dac_samples - 1;
+	const auto frames_end_index = frames_requested - 1;
+
+	static std::vector<AudioFrame> frames = {};
+
+	const auto is_zero_hold = (resample_method ==
+	                           ResampleMethod::ZeroOrderHoldAndResample);
+
+	auto stretch_sample = [&](float dac_index) {
+		const auto index_a  = std::max(0, static_cast<int>(dac_index));
+		const auto sample_a = static_cast<float>(dac_samples[index_a]);
+		const auto fraction = dac_index - static_cast<float>(index_a);
+
+		if (is_zero_hold || fraction == 0.0f) {
+			return sample_a;
+		} else {
+			const auto index_b  = std::min(index_a + 1, dac_end_index);
+			const auto sample_b = static_cast<float>(dac_samples[index_b]);
+			return lerp(sample_a, sample_b, fraction);
```

Comment:
Also, just realized.  This new function deviates in behavior from `main`.  The old code called this function instead of a call to `AddSamples` and did this simple resampling method in-place.  This new code does both.  So it will end up re-sampling a 2nd time (possibly going through Speex) and probably won't sound quite right.

Reply:
The original stretch code bypassed the sound blaster's own channel and was injecting samples directly into the final mix buffer, and needed to do it's own scaling and channel mapping. This is unlike any of the audio devices - even the LPT DACs, Tandy DAC, and PS1 DACs don't have this "direct feed" into the main mixer.

The new code is entirely self-contained inside the Sound Blaster. It resamples the DAC up to whatever the Sound Blaster's "hardware" output rate is set at (which may or may not be the SDL's mixer rate). This is entirely software controlled; games can set the SB to run at 22 KHz or 45 KHz.. and then they can still feed it DAC samples at some other rate (greater or less than).

So this new code goes from the DAC rate to the SB's own rate using the desired resampling method.  Typically the DAC rate is something like 8 KHz and the SB's running at 22 KHz. If ZOH is used, then you get the crunchy sound from 8 KHz -> 22 KHz, which I think, is 'proper'.

And from there, the SB's channel output ("crunchy 22 KHz") will be sampled again using ZOH up to whatever the user's mixer is running at.. typically 48 KHz.

Reply:
I agree with all those points in theory.  What's more important is testing and making sure it works well in practice.  SoundBlaster DAC gave me a bit of trouble at first.